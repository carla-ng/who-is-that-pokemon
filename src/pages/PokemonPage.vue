<template>
    <div id="app-inner-container">
        <h1 v-if="!pokemon">Please wait...</h1>

        <div v-else>
            <h1>Who is this Pok√©mon?</h1>
        
            <PokemonPicture
                :pokemonId="pokemon.id"
                :showPokemon="showPokemon"
            />

            <PokemonOptions
                :pokemons="pokemonArr"
                @pokemonSelection="checkAnswer"
            />

            <div v-if="showAnswer" class="fade-in">
                <h2>{{ message }}</h2>
                <button @click="newGame"> Play again </button>
            </div>
        </div>
    </div>
</template>


<script>
import PokemonPicture from '@/components/PokemonPicture'
import PokemonOptions from '@/components/PokemonOptions'

import getPokemonOptions from '@/helpers/getPokemonOptions'

export default {
    components: {
        PokemonPicture,
        PokemonOptions
    },
    data() {
        return {
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: '',
            disabledFlag: false
        }
    },
    methods: {
        async mixedPokemonArray() {
            this.pokemonArr = await getPokemonOptions()

            const randomInt = Math.floor ( Math.random() * 4 )
            this.pokemon = this.pokemonArr[ randomInt ]
        },
        checkAnswer( selectedId ) {
            this.showPokemon = true
            this.showAnswer  = true

            if (!this.disabledFlag) {
                this.disabledFlag = true
                let pokemonName = (this.pokemon.name).charAt(0).toUpperCase() + (this.pokemon.name).slice(1)

                if (selectedId == this.pokemon.id ) {
                    this.message = `Right answer! :) It's ${ pokemonName }`
                } else {
                    this.message = `Oops! Wrong answer :( It's ${ pokemonName }`
                }
            }
        },
        newGame() {
            this.showPokemon  = false
            this.showAnswer   = false
            this.pokemonArr   = []
            this.pokemon      = null
            this.disabledFlag = false

            this.mixedPokemonArray()
        }
    },
    mounted() {
        this.mixedPokemonArray()
    }
}
</script>

<style scoped>
    h1 {
        color: #000;
        font-family: Verdana, Geneva, sans-serif;
        letter-spacing: 2px;
        margin-bottom: 35px;
    }

    h2 {
        color: #000;
        font-family: Verdana, Geneva, sans-serif;
        font-size: 20px;
        letter-spacing: 2px;
        margin-top: 30px;
        padding: 0 12px
    }

    button {
        background-color: white;
        border-radius: 5px;
        border: 2px solid #000;
        color: #000;
        cursor: pointer;
        font-family: Verdana, Geneva, sans-serif;
        font-size: 16px;
        font-weight: 600;
        letter-spacing: 2px;
        margin: 10px 0 ;
        width: 200px;
        padding: 12px;
    }

    button:hover {
        background-color: rgba(56, 119, 255, 0.548);
    }
</style>
