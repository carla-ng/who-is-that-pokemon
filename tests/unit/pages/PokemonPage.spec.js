import { shallowMount, mount } from "@vue/test-utils"
import PokemonPage from "@/pages/PokemonPage"
import { pokemons } from "../mocks/pokemon.mock"


describe('PokemonPage Component', () => {

    let wrapper

    beforeEach(() => {
        wrapper = shallowMount( PokemonPage )
    });


    test('Must match snapshot', () => {
        expect( wrapper.html() ).toMatchSnapshot()
    })


    test('Must call mixedPokemonArray when mounting', () => {
        const mixedPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixedPokemonArray')
        const wrapper = shallowMount( PokemonPage )

        expect ( mixedPokemonArraySpy ).toHaveBeenCalled()
    })


    test('Must match snapshot after loading Pokemon', () => {
        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        expect( wrapper.html() ).toMatchSnapshot()
    })


    test('Must show PokemonPicture and PokemonOptions components', () => {
        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        const pokemonPicture = wrapper.find('pokemon-picture-stub')
        expect( pokemonPicture.exists() ).toBeTruthy()

        const pokemonOptions = wrapper.find('pokemon-options-stub')
        expect( pokemonOptions.exists() ).toBeTruthy()

        expect( pokemonPicture.attributes('pokemonid') ).toBe('1')
        expect( pokemonOptions.attributes('pokemons') ).toBeTruthy()

    })


    test('Testing checkAnswer', async() => {
        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        await wrapper.vm.checkAnswer(1)

        expect( wrapper.find('h2').exists() ).toBeTruthy()
        expect( wrapper.vm.showPokemon ).toBe(true)
        expect( wrapper.find('h2').text() ).toBe(`Right answer! :) It's ${ pokemons[0].name }`)

        await wrapper.vm.checkAnswer(2)
        expect( wrapper.vm.message ).toBe(`Oops! Wrong answer :( It's ${ pokemons[0].name }`)

    })

})