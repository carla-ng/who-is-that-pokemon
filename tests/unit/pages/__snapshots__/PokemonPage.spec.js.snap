// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component Must match snapshot 1`] = `<h1>Please wait...</h1>`;

exports[`PokemonPage Component Must match snapshot after loading Pokemon 1`] = `
<div>
  <h1>Who is this Pok√©mon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
